File: scripts/populate_db.js
import fs from 'fs-extra';
import path from 'path';
import crypto from 'crypto';
import { pipeline, env } from '@xenova/transformers';
import { ChromaClient } from 'chromadb';
import { v4 as uuidv4 } from 'uuid';

env.backend = 'onnxruntime';
env.allowRemoteModels = false;
env.localModelPath = '/home/rey/Documents/ai-doc-assit/model_convert/models/';


/**
 * Splits text into chunks of at most maxChunkSize characters.
 * Attempts to break on newlines or spaces.
 */
function chunkText(text, maxChunkSize = 1000) {
  text = text.replace(/\n/g, " ");
  text = text.split(/\s+/).join(" ");
  
  const chunks = [];
  let start = 0;
  while (start < text.length) {
    let end = start + maxChunkSize;
    if (end < text.length) {
      const lastNewline = text.lastIndexOf('\n', end);
      const lastSpace = text.lastIndexOf(' ', end);
      if (lastNewline > start) {
        end = lastNewline;
      } else if (lastSpace > start) {
        end = lastSpace;
      }
    }
    chunks.push(text.slice(start, end).trim());
    start = end;
  }
  return chunks;
}

/**
 * Compute MD5 hash of a given text.
 */
function computeMD5(text) {
  return crypto.createHash('md5').update(text, 'utf8').digest('hex');
}

/**
 * Process a file: read its content, split into chunks,
 * and return an array of chunk objects with metadata.
 */
function processFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  console.log(content);
  const chunks = chunkText(content, 1000); return chunks.map((chunk, index) => {
    const chunkId = uuidv4();
    const mdHash = computeMD5(chunk);
    const baseName = path.basename(filePath, path.extname(filePath));
    const chunkFilename = `${baseName}_${index}.txt`;
    return {
      id: chunkId,
      content: chunk,
      md_hash: mdHash,
      text_chunk_filename: chunkFilename,
      source_file: filePath,
    };
  });
}

/**
 * Loads all Markdown (or plain text) files from a directory,
 * processes them into chunks, and returns a flat array of chunk objects.
 */
async function loadAndProcessFiles(dir) {
  const files = await fs.readdir(dir);
  let allChunks = [];
  for (const file of files) {
    if (file.endsWith('.md') || file.endsWith('.txt')) {
      const filePath = path.join(dir, file);
      const chunks = processFile(filePath);
      allChunks = allChunks.concat(chunks);
    }
  }
  return allChunks;
}

async function main() {
  const dataDir = path.resolve('data/processed');
  const chunks = await loadAndProcessFiles(dataDir);
  console.log(`Processed ${chunks.length} chunks from files in ${dataDir}`);

  console.log("Loading local embedding model...");
  const embedder = await pipeline('feature-extraction', 'mixedbread-ai/mxbai-embed-xsmall-v1');

  for (const chunk of chunks) {
    const embedding = await embedder(chunk.content, { pooling: 'mean', normalize: true });
    chunk.embedding = Array.from(embedding.data);
  }
  console.log("Embeddings generated for all chunks.");

  console.log("Connecting to local ChromaDB...");
  const chroma = new ChromaClient({ path: "http://localhost:8000" });

  let collection;
  try {
    collection = await chroma.getCollection({ name: "docs" });
    console.log("Using existing collection 'docs'.");
  } catch (err) {
    collection = await chroma.createCollection({ name: "docs" });
    console.log("Created collection 'docs'.");
  }

  const ids = chunks.map(c => c.id);
  const documents = chunks.map(c => c.content);
  const embeddings = chunks.map(c => c.embedding);
  const metadatas = chunks.map(c => ({
    text_chunk_filename: c.text_chunk_filename,
    md_hash: c.md_hash,
    source_file: c.source_file,
  }));

  console.log("Adding chunks to the collection...");
  console.log({
    ids,
    documents,
    embeddings,
    metadatas,
  })

  console.log("Sample ID:", ids[0]);
console.log("Sample Document:", documents[0]);
console.log("Sample Embedding (length):", embeddings[0].length);
console.log("Sample Metadata:", metadatas[0]);


  await collection.add({
    ids,
    documents,
    embeddings,
    metadatas,
  });
  console.log("Successfully populated ChromaDB with document chunks.");
}

main().catch(err => {
  console.error("Error populating ChromaDB:", err);
});

File: scripts/populate_db.py
# https://stackoverflow.com/questions/76958817/streamlit-your-system-has-an-unsupported-version-of-sqlite3-chroma-requires-sq
__import__('pysqlite3')
import sys
sys.modules['sqlite3'] = sys.modules.pop('pysqlite3')

import os
import hashlib
import uuid
from pathlib import Path
from tqdm import tqdm
from sentence_transformers import SentenceTransformer
import chromadb
from chromadb.config import Settings

def chunk_text(text, max_chunk_size=1000):
    """Splits text into chunks of at most max_chunk_size characters.
    Tries to break on newlines or spaces for graceful chunking."""

    chunks = []
    start = 0
    while start < len(text):
        end = start + max_chunk_size
        if end < len(text):
            # try to break at the last newline, otherwise at the last space
            last_newline = text.rfind("\n", start, end)
            last_space = text.rfind(" ", start, end)
            if last_newline > start:
                end = last_newline
            elif last_space > start:
                end = last_space
        chunk = text[start:end].strip()
        if chunk:
            chunks.append(chunk)
        start = end
    return chunks

def compute_md5(text):
    """Compute MD5 hash for a given text string."""
    return hashlib.md5(text.encode("utf-8")).hexdigest()

def process_file(file_path):
    """Reads file content, splits into chunks, and returns a list of chunk dicts."""
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    chunks = chunk_text(content, max_chunk_size=1000)
    base_name = Path(file_path).stem
    chunk_objects = []
    for index, chunk in enumerate(chunks):
        chunk_id = str(uuid.uuid4())
        md_hash = compute_md5(chunk)
        chunk_filename = f"{base_name}_{index}.txt"
        chunk_objects.append({
            "id": chunk_id,
            "content": chunk,
            "md_hash": md_hash,
            "text_chunk_filename": chunk_filename,
            "source_file": str(file_path)
        })
    return chunk_objects

def load_and_process_files(directory):
    """Loads all .md and .txt files from directory and returns all chunk objects."""
    all_chunks = []
    for file in os.listdir(directory):
        if file.endswith(".md") or file.endswith(".txt"):
            file_path = os.path.join(directory, file)
            file_chunks = process_file(file_path)
            all_chunks.extend(file_chunks)
    return all_chunks

def main():
    # Define directory containing your plain text (or processed Markdown) files.
    data_dir = os.path.abspath("data/processed")
    chunks = load_and_process_files(data_dir)
    print(f"Processed {len(chunks)} chunks from files in {data_dir}")
    
    # Load the embedding model.
    # Here, we use mixedbread-ai/mxbai-embed-xsmall-v1 as an example.
    # (If this model isn't supported by SentenceTransformer, you might try an alternative.)
    print("Loading local embedding model...")
    # model = SentenceTransformer("mixedbread-ai/mxbai-embed-xsmall-v1")
    model = SentenceTransformer("Xenova/jina-embeddings-v2-base-en")
    
    # Generate embeddings for each chunk.
    for chunk in tqdm(chunks, desc="Generating embeddings"):
        # Pool with 'mean' and normalize the vector.
        embedding = model.encode(chunk["content"], show_progress_bar=False, normalize_embeddings=True)
        # Convert embedding (if numpy array) to list.
        chunk["embedding"] = embedding.tolist() if hasattr(embedding, "tolist") else embedding
    
    print("Embeddings generated for all chunks.")
    
    # Connect to local ChromaDB instance (ensure it is running, e.g., via Docker on http://localhost:8000)
    print("Connecting to local ChromaDB...")
    # client = chromadb.PersistentClient(
    #     settings=Settings(chroma_db_impl="duckdb+parquet", persist_directory="chroma_db")
    # )
    client = chromadb.HttpClient(host='localhost', port=8000)
    
    # Get or create a collection named "docs"
    try:
        collection = client.get_collection(name="docs")
        print("Using existing collection 'docs'.")
    except Exception as e:
        collection = client.create_collection(name="docs")
        print("Created collection 'docs'.")
    
    # Prepare arrays for ChromaDB.
    ids = [chunk["id"] for chunk in chunks]
    documents = [chunk["content"] for chunk in chunks]
    embeddings = [chunk["embedding"] for chunk in chunks]
    metadatas = [{
        "text_chunk_filename": chunk["text_chunk_filename"],
        "md_hash": chunk["md_hash"],
        "source_file": chunk["source_file"]
    } for chunk in chunks]
    
    # Add all chunks to the collection.
    print("Adding chunks to the collection...")
    # print(embeddings)
    collection.add(ids=ids, documents=documents, embeddings=embeddings, metadatas=metadatas)
    print("Successfully populated ChromaDB with document chunks.")

if __name__ == "__main__":
    main()
